code/parse/sexp/EngineSEXP.h

```diff
@@ -0,0 +1,198 @@
+#pragma once
+
+#include "globalincs/pstypes.h"
+
+#include "DynamicSEXP.h"
+#include "SEXPParameterExtractor.h"
+
+namespace sexp {
+
+class EngineSEXP;
+
+struct dummy_return {
+};
+
+using EngineSexpAction = std::function<int(SEXPParameterExtractor* extractor)>;
+
+class EngineSEXPFactory {
+	std::unique_ptr<EngineSEXP> m_sexp;
+
+	SCP_string _helpText;
+
+	int _category = -1;
+
+	int _subcategory = -1;
+	SCP_string _subcategoryName;
+
+	int _returnType = -1;
+
+	EngineSexpAction _action;
+
+	struct argument {
+		int type = -1;
+		SCP_string help_text;
+
+		bool optional_marker = false;
+		bool varargs_marker = false;
+	};
+
+	SCP_vector<argument> _arguments;
+
+	class ArgumentListBuilder {
+		EngineSEXPFactory* _parent;
+
+	  public:
+		ArgumentListBuilder(EngineSEXPFactory* parent);
+
+		/**
+		 * @brief Adds a parameter of the specified type with help text.
+		 * @param type The type of this parameter
+		 * @param help_text A helpful text for this parameter
+		 * @return The arglist builder
+		 */
+		ArgumentListBuilder& arg(int type, SCP_string help_text);
```

Comment:
Should this be `const SCP_string &`?

Reply:
No, that is intentional since the string is `std::move`d inside and this avoid a copy operation.